torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -587,43 +595,19 @@ def forward(self, hidden_states):
         return y
 
     def moe_forward(self, x, topk_ids, topk_weight):
-        # This part sorts the token indices so that tokens routed to the same expert reside consecutively.
-        # An implication is that tokens to the same "expert group" (i.e., device) are also consecutive.
-        # Since this is an "aritificial" index creation (final outcome being
-        # `idxs`), we don't need gradients here.
-        with torch.no_grad():
-            # [seq_len, n_routed_experts]
-            cnts = topk_ids.new_zeros((topk_ids.shape[0], self.config.n_routed_experts))
-            # Fill 1 to the selected experts
-            cnts.scatter_(1, topk_ids, 1)
-            tokens_per_expert = cnts.sum(dim=0)
-            # Token indices for each expert
-            idxs = topk_ids.view(-1).argsort()
-            sorted_tokens_shape = idxs.shape + x.shape[1:]
-
-        sorted_tokens = x[idxs // topk_ids.shape[1]]
-        assert sorted_tokens.shape == sorted_tokens_shape
-
-        # This part exchange the information about the number of tokens send and
-        # received by each expert. We can understand this information as "side
-        # band", which is not part of the actual data. Thus no gradient is
-        # needed.
-        with torch.no_grad():
-            # Sum the tokens over local experts, then we get tokens per EP rank,
-            # which is the input splits
-            tokens_per_expert_group = tokens_per_expert.new_empty(
-                tokens_per_expert.shape[0]
-            )
-            dist.all_to_all_single(
-                tokens_per_expert_group, tokens_per_expert, group=self.ep_group
-            )
-            input_splits = tokens_per_expert.view(self.ep_size, -1).sum(dim=1)
+        (
+            sorted_tokens,
+            token_indices,
+            sorted_tokens_shape,
+            input_splits,
+            tokens_per_expert_group,
+        ) = self.prepare_expert_routing(x, topk_ids, topk_weight)
```

Comment:
While consolidating code is great, this signature looks a bit chunky, and opaque.
I wonder if there is a mid point. 
Maybe we can make a `sort_tokens` method and leave the all-to-all alone? (It is sometimes okay to have duplicated code :) )

Reply:
makes sense. I actually thought about splitting this into two functions earlier. 
Anyway, for now have made the lighter 'sort tokens' and left the allto all in duplicate.